### ISSUE TITLE
defined(${'!'}) versus defined(${'::!'})

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#11594]
Migrated from [rt.perl.org#97492](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=97492) (status was 'resolved')

Searchable as RT97492$

**Comment by p5pRT** [comment]
### From @cpansprout

Perl pretends that it’s built\-in vars all exist\, even though it actually creates most of them on the fly\. I’m referring to the behaviour of defined\(\)\. It \*will\* create built\-in variables\, even though it doesn’t create other variables\, but leaves them nonexistent\.

However\, the logic in gv\.c&#8203;:is\_gv\_magical\_sv is faulty\, because it does not take package prefixes into account&#8203;:

$ \./miniperl \-le 'print defined $\{q"\!"\}'
\! at \-e line 1\.
1
$ \./miniperl \-le 'print defined $\{q"&#8203;::\!"\}'
:&#8203;:\! at \-e line 1\.



---
Flags&#8203;:
&nbsp;    category=core
&nbsp;    severity=low


---
Site configuration information for perl 5\.15\.1&#8203;:

Configured by sprout at Fri Jul 22 19&#8203;:53&#8203;:36 PDT 2011\.

Summary of my perl5 \(revision 5 version 15 subversion 1\) configuration&#8203;:
&nbsp;  Snapshot of&#8203;: 862b2c434beb1d61a19037a8449c8db953fd37a0
&nbsp;  Platform&#8203;:
&nbsp;    osname=darwin\, osvers=10\.5\.0\, archname=darwin\-2level
&nbsp;    uname='darwin pint\.local 10\.5\.0 darwin kernel version 10\.5\.0&#8203;: fri nov 5 23&#8203;:20&#8203;:39 pdt 2010; root&#8203;:xnu\-1504\.9\.17~1release\_i386 i386 '
&nbsp;    config\_args='\-de \-Dusedevel'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    useithreads=undef\, usemultiplicity=undef
&nbsp;    useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
&nbsp;    use64bitint=undef\, use64bitall=undef\, uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-fno\-common \-DPERL\_DARWIN \-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include'\,
&nbsp;    optimize='\-O3'\,
&nbsp;    cppflags='\-fno\-common \-DPERL\_DARWIN \-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='4\.2\.1 \(Apple Inc\. build 5664\)'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=8\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='env MACOSX\_DEPLOYMENT\_TARGET=10\.3 cc'\, ldflags =' \-fstack\-protector \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /usr/lib
&nbsp;    libs=\-ldbm \-ldl \-lm \-lutil \-lc
&nbsp;    perllibs=\-ldl \-lm \-lutil \-lc
&nbsp;    libc=\, so=dylib\, useshrplib=false\, libperl=libperl\.a
&nbsp;    gnulibc\_version=''
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=bundle\, d\_dlsymun=undef\, ccdlflags=' '
&nbsp;    cccdlflags=' '\, lddlflags=' \-bundle \-undefined dynamic\_lookup \-L/usr/local/lib \-fstack\-protector'

Locally applied patches&#8203;:
&nbsp;    



---
@&#8203;INC for perl 5\.15\.1&#8203;:
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.15\.1/darwin\-2level
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.15\.1
&nbsp;    /usr/local/lib/perl5/5\.15\.1/darwin\-2level
&nbsp;    /usr/local/lib/perl5/5\.15\.1
&nbsp;    /usr/local/lib/perl5/site\_perl
&nbsp;    \.



---
Environment for perl 5\.15\.1&#8203;:
&nbsp;    DYLD\_LIBRARY\_PATH \(unset\)
&nbsp;    HOME=/Users/sprout
&nbsp;    LANG=en\_US\.UTF\-8
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=/usr/bin&#8203;:/bin&#8203;:/usr/sbin&#8203;:/sbin&#8203;:/usr/local/bin&#8203;:/usr/X11/bin&#8203;:/usr/local/bin
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    SHELL=/bin/bash

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Aug 21 20&#8203;:34&#8203;:34 2011\, sprout wrote&#8203;:
> Perl pretends that it’s built\-in vars all exist\, even though it
>    actually creates most of them on the fly\. I’m referring to the
>    behaviour of defined\(\)\. It \*will\* create built\-in variables\, even
>    though it doesn’t create other variables\, but leaves them
>    nonexistent\.
> 
> However\, the logic in gv\.c&#8203;:is\_gv\_magical\_sv is faulty\, because it does
>    not take package prefixes into account&#8203;:
> 
> $ \./miniperl \-le 'print defined $\{q"\!"\}'
> \! at \-e line 1\.
> 1
> $ \./miniperl \-le 'print defined $\{q"&#8203;::\!"\}'
> :&#8203;:\! at \-e line 1\.
>

Er\, ignore the first line of output from each command\.  That was my
tweaked build with a Perl\_warn\(\) statement added\.  It should be like this&#8203;:

$ \./miniperl \-le 'print defined $\{q"\!"\}'
1
$ \./miniperl \-le 'print defined $\{q"&#8203;::\!"\}'

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Sun Aug 21 20&#8203;:34&#8203;:34 2011\, sprout wrote&#8203;:
> Perl pretends that it’s built\-in vars all exist\, even though it
>    actually creates most of them on the fly\. I’m referring to the
>    behaviour of defined\(\)\. It \*will\* create built\-in variables\, even
>    though it doesn’t create other variables\, but leaves them
>    nonexistent\.
> 
> However\, the logic in gv\.c&#8203;:is\_gv\_magical\_sv is faulty\, because it does
>    not take package prefixes into account&#8203;:
> 
> $ \./miniperl \-le 'print defined $\{q"\!"\}'
> \! at \-e line 1\.
> 1
> $ \./miniperl \-le 'print defined $\{q"&#8203;::\!"\}'
> :&#8203;:\! at \-e line 1\.
>

Er\, ignore the first line of output from each command\.  That was my
tweaked build with a Perl\_warn\(\) statement added\.  It should be like this&#8203;:

$ \./miniperl \-le 'print defined $\{q"\!"\}'
1
$ \./miniperl \-le 'print defined $\{q"&#8203;::\!"\}'

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @Hugmeir

On Mon\, Aug 22\, 2011 at 12&#8203;:34 AM\, Father Chrysostomos \<
perlbug\-followup@&#8203;perl\.org> wrote&#8203;:

>
> However\, the logic in gv\.c&#8203;:is\_gv\_magical\_sv is faulty\, because it does not
> take package prefixes into account&#8203;:
>

I can't seem to find the tests for defined\(\)\, but this patch should make do
for the report\, unless I'm missing some weird prefixing behavior\.
\(Also\, that function isn't nul\-clean\, but I have that taken care of in the
GSOC repo\.\)

<details><summary>Inline Patch</summary>

```diff
diff --git a/gv.c b/gv.c
index 228ab45..d7d1c06 100644
--- a/gv.c
+++ b/gv.c
@@ -2800,11 +2800,27 @@ bool
 Perl_is_gv_magical_sv(pTHX_ SV *const name_sv, U32 flags)
 {
     STRLEN len;
-    const char *const name = SvPV_const(name_sv, len);
+    const char *name = SvPV_const(name_sv, len);

     PERL_UNUSED_ARG(flags);
     PERL_ARGS_ASSERT_IS_GV_MAGICAL_SV;

+    if ( len > 1 ) {
+        if (name[0] == '\'' && name[1] != '\0') {
+            name++;
+            len--;
+        }
+        else if (len > 2 && name[0] == ':' && name[1] == ':') {
+            name += 2;
+            len  -= 2;
+        }
+
+        while ( len > 6 && name[0] == 'm' && memEQ(name, 6, "main::") ) {
+            name += 6;
+            len  -= 6;
+        }
+    }
+
     if (len > 1) {
     const char * const name1 = name + 1;
     switch (*name) {
```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

On Tue Aug 23 21&#8203;:09&#8203;:04 2011\, fraserbn@&#8203;gmail\.com wrote&#8203;:
> On Mon\, Aug 22\, 2011 at 12&#8203;:34 AM\, Father Chrysostomos \<
> perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> 
> >
> > However\, the logic in gv\.c&#8203;:is\_gv\_magical\_sv is faulty\, because it
does not
> > take package prefixes into account&#8203;:
> >
> 
> I can't seem to find the tests for defined\(\)\, but this patch should
make do
> for the report\, unless I'm missing some weird prefixing behavior\.

\*gonzo = \*main&#8203;::;
print defined \*\{"gonzo&#8203;::\."\};

Maybe fix it it like this&#8203;:
1\. look for the last package separator
2\. if there is one\, fetch the package \(taking care of s/'\\z/&#8203;::/\)
3\. if it’s not PL\_defstash\, return false
4\. proceed as usual

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Tue Aug 23 21&#8203;:09&#8203;:04 2011\, fraserbn@&#8203;gmail\.com wrote&#8203;:
> On Mon\, Aug 22\, 2011 at 12&#8203;:34 AM\, Father Chrysostomos \<
> perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> 
> >
> > However\, the logic in gv\.c&#8203;:is\_gv\_magical\_sv is faulty\, because it
does not
> > take package prefixes into account&#8203;:
> >
> 
> I can't seem to find the tests for defined\(\)\, but this patch should
make do
> for the report\, unless I'm missing some weird prefixing behavior\.

\*gonzo = \*main&#8203;::;
print defined \*\{"gonzo&#8203;::\."\};

Maybe fix it it like this&#8203;:
1\. look for the last package separator
2\. if there is one\, fetch the package \(taking care of s/'\\z/&#8203;::/\)
3\. if it’s not PL\_defstash\, return false
4\. proceed as usual

**Comment by p5pRT** [comment]
### From @cpansprout

This was fixed in commit 23496c6e\.  Tests were added in commit b4155db\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
